libflux/src/flux/ast/mod.rs

```diff
@@ -367,6 +385,8 @@ pub struct File {
     #[serde(deserialize_with = "deserialize_default_from_null")]
     pub imports: Vec<ImportDeclaration>,
     pub body: Vec<Statement>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub eof: CommentList,
```

Comment:
So if I understand correctly, comments are associated with the trailing token/node, but in the case of comments at the end of the file there is no token so we simple capture them here as part of the File node?

So if I am walking the AST looking for comments I can expect to find them in the BaseNode as part of each node  or I will find comments inside various other nodes like the File and Block nodes to capture comments that do not have a clear association with a token/node. Does that sound like I have it right?

Reply:
That's right, the comments at the end of the file are a special case because there is no token in front to associate them with. You can think of it as preserving the EOF token.

In Colm, comments and other ignored things can be attached to the front and back of tokens. It requires altering the scanner quite a bit. After scanning a token you have to scan ahead only comments and attach them to the token, before you send the token off. I actually tried this approach, but kept running into some strange behavior that I couldn't explain. But we decided to go ahead with just comments in front so I dropped that investigation.

For our case, comments after tokens is mostly overkill, with one exception:

let a = 1; // set a to one

It's nice to have comments at the end of a line attached to the end of the last token. For this we will need special case added to the formatter. It's currently not smart enough to preserve this. I don't do much commenting that way, but I believe lots of people do.